pkg/metrics/collector/pod_ip_metrics.go

```diff
@@ -0,0 +1,174 @@
+/*
+Copyright 2022 Google Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package collector
+
+import (
+	"bufio"
+	"fmt"
+	"net"
+	"os"
+	"strings"
+
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type ipFamily int
+
+const (
+	gkePodNetworkDir = "/host/var/lib/cni/networks/gke-pod-network"
+	ipv4 ipFamily = 0
+	ipv6 = 1
+	dual = 2
+	dualStack = "IPV4_IPV6"
+)
+
+var (
+	usedIpv4AddrCountDesc = prometheus.NewDesc(
+		"used_ipv4_addr_count",
+		"Indicates how many IPv4 addresses are in use.",
+		nil, nil,
+	)
+	usedIpv6AddrCountDesc = prometheus.NewDesc(
+		"used_ipv6_addr_count",
+		"Indicates how many IPv6 addresses are in use.",
+		nil, nil,
+	)
+	dualStackCountDesc = prometheus.NewDesc(
+		"dual_stack_count",
+		"Indicates how many pods have dual stack IP addresses.",
+		nil, nil,
+	)
+	dualStackErrorCountDesc = prometheus.NewDesc(
+		"dual_stack_error_count",
+		"Indicates how many pods did not get dual stack IPs erroneously.",
+		nil, nil,
+	)
+	duplicateIpCountDesc = prometheus.NewDesc(
+		"duplicate_ip_count",
+		"Indicates how many pods had more than one IP per family.",
+		nil, nil,
+	)
+)
+
+type podIpMetricsCollector struct {
+	usedIpv4AddrCount uint64
+	usedIpv6AddrCount uint64
+	dualStackCount uint64
+	dualStackErrorCount uint64
+	duplicateIpCount uint64
+}
+
+func init() {
+	registerCollector("pod_ip_metrics", NewPodIpMetricsCollector)
+}
+
+// NewPodIpMetricsCollector returns a new Collector exposing pod IP allocation
+// stats.
+func NewPodIpMetricsCollector() (Collector, error) {
+	return &podIpMetricsCollector{}, nil
+}
+
+func readLine(path string) (string, error) {
+	buf, err := os.Open(path)
+	if err != nil {
+		return "", err
+	}
+
+	defer func() {
+		if err = buf.Close(); err != nil {
+			glog.Errorf("Error closing file %s: %s", path, err)
+		}
+	}()
+
+	s := bufio.NewScanner(buf)
+	s.Scan()
+	return s.Text(), s.Err()
+}
+
+func (c *podIpMetricsCollector) listIpAddresses(dir string) error {
+	f, err := os.Open(dir)
+	if err != nil {
+		glog.Errorf("Error opening directory %s, %v", dir, err)
```

Comment:
Just wanted to point out there may be a short period of time when the `gke-pod-network` folder doesn't exist yet (e.g. no pod has been created or only hostnetwork pods are created). That however should be fine as we only log an error and retry later.

Reply:
Yes. I noticed that and this is ok.